### PR TITLE
port clerk testcases

### PR DESCRIPTION
**Pull request by temaniarpit27**

### CODE REVIEW

File: x/clerk/module_test.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package clerk_test
+
```
Comment (ssandeep) [inline x/clerk/module_test.go:2]: is this file needed?

File: x/clerk/keeper/keeper_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+func (suite *KeeperTestSuite) TestSetHasGetRecordSequence() {
+	t, app, ctx := suite.T(), suite.app, suite.ctx
+
+	testSeq := "testseq"
+	ck := app.ClerkKeeper
+	ck.SetRecordSequence(ctx, testSeq)
+	found := ck.HasRecordSequence(ctx, testSeq)
+	require.True(t, found)
+
+	found = ck.HasRecordSequence(ctx, "testSeq")
```
Comment (ssandeep) [inline x/clerk/keeper/keeper_test.go:147]: Did you mean to use the variable instead?

File: x/clerk/keeper/msg_server.go [hunk 3]
```diff
@@ -45,15 +45,16 @@ func (k msgServer) MsgEventRecord(goCtx context.Context, msg *types.MsgEventReco
...
 
 	// check chain id
-	if chainParams.BorChainID != msg.ChainId {
-		k.Logger(ctx).Error("Invalid Bor chain id", "msgChainID", msg.ChainId)
-		return nil, hmCommon.ErrInvalidBorChainID
-	}
+	// if chainParams.BorChainID != msg.ChainId {
+	// 	k.Logger(ctx).Error("Invalid Bor chain id", "msgChainID", msg.ChainId)
+	// 	return nil, hmCommon.ErrInvalidBorChainID
+	// }
```
Comment (ssandeep) [inline x/clerk/keeper/msg_server.go:57]: This TODO to be taken up later?

File: x/clerk/side_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	tmprototypes "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	// ethTypes "github.com/maticnetwork/bor/core/types"
+	// common "github.com/maticnetwork/bor/common"
```
Comment (ssandeep) [inline x/clerk/side_handler_test.go:17]: let's remove these lines

File: x/clerk/side_handler_test.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+	// 	// mock external calls -- no receipt
+	// 	suite.contractCaller.On("GetConfirmedTxReceipt", txHash.EthHash(), chainParams.MainchainTxConfirmations).Return(txReceipt, nil)
+	// 	suite.contractCaller.On("DecodeStateSyncedEvent", chainParams.ChainParams.StateSenderAddress, txReceipt, logIndex).Return(nil, nil)
+
+	// 	// execute handler
+	// 	result := suite.sideHandler(ctx, &msg)
+	// 	// require.NotEqual(t, uint32(sdk.CodeOK), result.Code, "Side tx handler should fail")
+	// 	require.Equal(t, tmprototypes.SideTxResultType_YES, result.Result, "Result should be `skip`")
+	// })
+}
```
Comment (ssandeep) [inline x/clerk/side_handler_test.go:185]: Any dependency on other modules?

File: x/clerk/side_handler_test.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+func (suite *SideHandlerTestSuite) TestPostHandler() {
+	t, ctx := suite.T(), suite.ctx
+
+	// post tx handler
+	_, err := suite.postHandler(ctx, nil, tmprototypes.SideTxResultType_YES)
+	require.Error(t, err)
+	// require.False(t, result.IsOK(), "Post handler should fail")
+	// require.Equal(t, sdk.CodeUnknownRequest, result.Code)
```
Comment (ssandeep) [inline x/clerk/side_handler_test.go:194]: Let's check if the exact error is returned and also for NotNil of the return value.

File: x/clerk/side_handler_test.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		suite.chainID,
+	)
+
+	t.Run("NoResult", func(t *testing.T) {
+		_, err := suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_NO)
+		require.Error(t, err)
+		// require.False(t, result.IsOK(), "Post handler should fail")
+		// require.Equal(t, common.CodeSideTxValidationFailed, result.Code)
+		// require.Equal(t, 0, len(result.Events), "No error should be emitted for failed post-tx")
+
```
Comment (ssandeep) [inline x/clerk/side_handler_test.go:224]: Same here as well.

File: x/clerk/side_handler_test.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+		)
+
+		_, err := suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_YES)
+		require.Nil(t, err)
+		// require.True(t, result.IsOK(), "Post handler should succeed")
+
+		_, err = suite.postHandler(ctx, &msg, tmprototypes.SideTxResultType_YES)
+		require.Error(t, err)
+		// require.False(t, result.IsOK(), "Post handler should prevent replay attack")
+		// require.Equal(t, common.CodeOldTx, result.Code)
```
Comment (ssandeep) [inline x/clerk/side_handler_test.go:277]: I feel we should check the following in all cases:
1. for error: (a) returned value should be nil, (b) error should be present, (c) error returned should be equated with the expected error
2. for success: (a) error should be nil, (b) returned value should be present, (c) if there's an predetermined return value, that should be equated with the actual value returned

**Review by ssandeep — CHANGES_REQUESTED** [review]
Let's also check `make lint`

File: x/clerk/handler_test.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+func (suite *HandlerTestSuite) SetupTest() {
+	suite.app, suite.ctx, _ = createTestApp(false)
+	suite.contractCaller = mocks.IContractCaller{}
+	suite.handler = clerk.NewHandler(suite.app.ClerkKeeper, &suite.contractCaller)
+
+	// TODO - Check this
+	// fetch chain id
+	// suite.chainID = suite.app.ChainKeeper.GetParams(suite.ctx).ChainParams.BorChainID
```
[the remark this answers is not in the document]
Comment (temaniarpit27) [reply x/clerk/handler_test.go:46]: @murthyvitwit any solution you got for this?

File: x/clerk/handler_test.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+				msg.Data,
+				msg.ChainId,
+				time.Now(),
+			),
+		)
+
+		_, err := suite.handler(ctx, &msg)
+		require.Error(t, err)
+		// require.False(t, result.IsOK(), "should fail due to existent event record but succeeded")
+		// require.Equal(t, types.CodeEventRecordAlreadySynced, result.Code)
```
[the remark this answers is not in the document]
Comment (temaniarpit27) [reply x/clerk/handler_test.go:113]: done
checking for result

File: x/clerk/handler_test.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+	// sequence id
+	blockNumber := new(big.Int).SetUint64(msg.BlockNumber)
+	sequence := new(big.Int).Mul(blockNumber, big.NewInt(hmTypes.DefaultLogIndexUnit))
+	sequence.Add(sequence, new(big.Int).SetUint64(msg.LogIndex))
+	app.ClerkKeeper.SetRecordSequence(ctx, sequence.String())
+
+	_, err := suite.handler(ctx, &msg)
+	require.Error(t, err)
+	// require.False(t, result.IsOK(), "should fail due to existent sequence but succeeded")
+	// require.Equal(t, common.CodeOldTx, result.Code)
```
[the remark this answers is not in the document]
Comment (temaniarpit27) [reply x/clerk/handler_test.go:142]: done

File: x/clerk/handler_test.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+// TODO - Check this
```
[the remark this answers is not in the document]
Comment (temaniarpit27) [reply x/clerk/handler_test.go:145]: dependent upon code commented above (GetParams)

File: x/clerk/module_test.go (hunk 1 above)
Comment (temaniarpit27) [reply x/clerk/module_test.go:2]: removed

File: x/clerk/keeper/msg_server.go (hunk 3 above)
Comment (temaniarpit27) [reply x/clerk/keeper/msg_server.go:57]: yes

File: x/clerk/keeper/keeper_test.go (hunk 2 above)
Comment (temaniarpit27) [reply x/clerk/keeper/keeper_test.go:147]: no
this is different

File: x/clerk/side_handler_test.go (hunk 4 above)
Comment (temaniarpit27) [reply x/clerk/side_handler_test.go:17]: not removing as these will be used when some tests will be uncommented

File: x/clerk/side_handler_test.go (hunk 5 above)
Comment (temaniarpit27) [reply x/clerk/side_handler_test.go:185]: based on chain params which is not working
needed @murthyvitwit help

File: x/clerk/side_handler_test.go (hunk 8 above)
Comment (temaniarpit27) [reply x/clerk/side_handler_test.go:277]: @ssandeep : any idea how to check for exact error?

Comment (ssandeep) [reply x/clerk/side_handler_test.go]: You can refer https://github.com/maticnetwork/heimdall/pull/519/files#diff-2e0ea1a4c73ff1bbc13d1079e303011edb4c62cd5e5533d859a667f1ceff8b9eR111

File: x/clerk/keeper/common_test.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
+package keeper_test
```
Comment (ssandeep) [inline x/clerk/keeper/common_test.go:1]: We can deduplicate the code in common_test inside clerk/keeper/ and clerk/
@saikumar0x1967 has done it in staking and checkpoint. Let's follow that.

**Comment by temaniarpit27** [comment]
@ssandeep fixed review comments

**Review by ssandeep — APPROVED** [review]
